common/src/stack/command/stack/commands/report/host/repo/imp_yum.py

```diff
@@ -0,0 +1,68 @@
+# @copyright@
+# Copyright (c) 2006 - 2019 Teradata
+# All rights reserved. Stacki(r) v5.x stacki.com
+# https://github.com/Teradata/stacki/blob/master/LICENSE.txt
+# @copyright@
+
+import jinja2
+
+import stack.commands
+from stack.repo import build_repo_files, yum_repo_template
+
+class Implementation(stack.commands.Implementation):
+
+	def run(self, args):
+		host	= args[0]
+		osname	= self.owner.host_attrs[host]['os']
+		box	= self.owner.host_attrs[host]['box']
+
+		if osname == 'sles':
+			repo_location = '/etc/zypp/repos.d/stacki.repo'
+		elif osname == 'redhat':
+			repo_location = '/etc/yum.repos.d/stacki.repo'
+
+		# if we ever decide to make the repo server ip address more abstract, this is where we'd change it:
+		stacki_http_url_base = 'http://{{ Kickstart_PrivateAddress }}/install'
+
+		repo_data = {}
+		# convert the raw pallet and cart data into repo dictionaries
+		for pallet in self.owner.box_data[box]['pallets']:
+			repo_data[pallet.name] = {
+				'alias': f'{pallet.name}-{pallet.version}-{pallet.rel}',
+				'name': f'{pallet.name} {pallet.version} {pallet.rel}',
+				'url': f'{stacki_http_url_base}/pallets/{pallet.name}/{pallet.version}/{pallet.rel}/{pallet.os}/{pallet.arch}',
+				'gpgcheck': False,
+			}
+
+		for cart, _ in self.owner.box_data[box]['carts']:
+			cart_url_base = 'http://{{ Kickstart_PrivateAddress }}/install/carts'
+			repo_data[f'{cart}-cart'] = {
+				'alias': f'{cart}-cart',
+				'name': f'{cart} cart',
+				'url': f'{stacki_http_url_base}/carts/{cart}',
+				'gpgcheck': False,
+			}
+
+		# repo objects are already in the right format so merge that dict here.
+		repo_data = {**repo_data, **self.owner.box_data[box]['repos']}
+
+		# build the template string lines from all the repo objects, add an empty line between each repo
+		repo_lines = '\n\n'.join(build_repo_files({box: repo_data}, yum_repo_template))
+
+		# make a second jinja pass at the repo data, to render any stacki attributes used (eg. Kickstart_PrivateAddress)
+		repo_str = jinja2.Template(repo_lines).render(**self.owner.host_attrs[host])
+
+		repo = []
+		repo.append('<stack:file stack:name="%s">' % repo_location)
+		repo.extend(repo_str.splitlines())
+		repo.append('</stack:file>')
+
+		if osname == 'sles':
```

Comment:
It would lead to duplicated code but would it be more clear if there was an `imp_zypper.py` vs a conditional for zypper here?

Reply:
I disagree with 🧠 . We have if statements for a reason. IMP should only be used when the core logic is significantly different. Less chance of code diverging over time, having to fix the same bug in multiple places, etc.

Reply:
I don't think I agree.

That said, maybe there's an argument here for something in pylib that hides details like that.

Picture an `OS` class that knows these differences.  Then, at that line, rather than a conditional, `OS(os_name).package_manager_cache_cleaner` returns `'zypper clean --all'` when `os_name == 'sles'`.  `OS(os_name).package_manager_confd` returns `/etc/zypp/repos.d/`

@bgreenb thoughts?

@caladd how bad an idea on a scale from 1-10?

Reply:
I'd implement it to be transparent, create a config module with a SLES dataclass, and REDHAT dataclass, and then at the top of the file have some logic to map an OS variable that is assigned to either SLES or REDHAT depending on the OS it is running on. Then everywhere else you would do `from config import OS` and have everything pull from that.

But, unless we are going to refactor all our code to use some sort of config system like that, I think it is a layer of abstraction that isn't needed. The only benefit would be gathering all the OS specific hard coded stuff in one place, to make it easier to support new OSes. Most the config strings wouldn't be used in multiple places, and you are just engineering an extra layer of complexity.

I say KISS (Keep it simple sweetie).

Reply:
Yeah, grepping quickly around our code, I'm not sure how much it would help.  Sometimes we just need to do wholly different things for sles vs redhat.  If we ignore those, we end up with a few more places like the above line that started the conversation, but they're usually trivial enough.

Reply:
@bsanders @caladd Alright that answers my question (sorry if it came off like a suggestion), whatever slight confusion (to me at least) of the implementation file name isn't worth the tradeoff of more code complexity. As to your idea, It sounds like it wouldn't add that much as it seems less easy to read than just calling the package manager command.